keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md

```diff
@@ -0,0 +1,249 @@
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [Stable](#stable)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This proposal aims to improve the Container Runtime Interface (CRI) APIs for
+managing a Container's CPU and memory resource configurations on the runtime.
+It seeks to extend UpdateContainerResources CRI API such that it works for
+Windows, and other future runtimes besides Linux. It also seeks to extend
+ContainerStatus CRI API to allow Kubelet to discover the current resources
+configured on a Container.
+
+## Motivation
+
+In-Place Pod Vertical Scaling feature relies on Container Runtime Interface
+(CRI) to update the CPU and/or memory limits for Container(s) in a Pod.
+
+The current CRI API set has a few drawbacks that need to be addressed:
+1. UpdateContainerResources CRI API takes a parameter that describes Container
+   resources to update for Linux Containers, and this may not work for Windows
+   Containers or other potential non-Linux runtimes in the future.
+1. There is no CRI mechanism that lets Kubelet query and discover the CPU and
+   memory limits configured on a Container from the Container runtime.
+1. The expected behavior from a runtime that handles UpdateContainerResources
+   CRI API is not very well defined or documented.
+
+### Goals
+
+This proposal has two primary goals:
+  - Modify UpdateContainerResources to allow it to work for Windows Containers,
+    as well as Containers managed by other runtimes besides Linux,
+  - Provide CRI API mechanism to query the Container runtime for CPU and memory
+    resource configurations that are currently applied to a Container.
+
+An additional goal of this proposal is to better define and document the
+expected behavior of a Container runtime when handling resource updates.
+
+### Non-Goals
+
+Definition of expected behavior of a Container runtime when it handles CRI APIs
+related to a Container's resources is intended to be a high level guide.  It is
+a non-goal of this proposal to define a detailed or specific way to implement
+these functions. Implementation specifics are left to the runtime, within the
+bounds of expected behavior.
+
+## Proposal
+
+One key change is to make UpdateContainerResources API work for Windows, and
+any other future runtimes, besides Linux by making the resources parameter
+passed in the API specific to the target runtime.
+
+Another change in this proposal is to extend ContainerStatus CRI API such that
+Kubelet can query and discover the CPU and memory resources that are presently
+applied to a Container.
+
+To accomplish aforementioned goals:
+
+* A new protobuf message object named *ContainerResources* that encapsulates
+LinuxContainerResources and WindowsContainerResources is introduced as below.
+  - This message can easily be extended for future runtimes by simply adding a
+    new runtime-specific resources struct to the ContainerResources message.
+```
```

Comment:
The CRI change looks good to me. 

@Random-Liu you left some comments earlier. Are your comments resolved?

Reply:
@yujuhong Thanks for the review. Yes all his comments have been addressed. Regarding deprecating use of UpdateContainerResourcesRequest.Linux, we don't have exact details but it should be a manageable task given the limited surface area.  We should be able to use [deprecated=true](https://developers.google.com/protocol-buffers/docs/proto3) strategy. IMHO, it isn't a blocking issue.